### DISCUSSION TITLE
Basic auth host rule setting not used for dependency lookup in private npm nexus repository

### DISCUSSION THREAD

**Discussion by luk-lo** [renovatebot/renovate#14131] [Request Help] [answered]
Hi, I'm using a renovate instance hosted by my organization and trying to configure access to a private npm nexus repository for the dependency version lookup step.
The relevant part of the configuration looks like this:
```json
"encrypted": {
  "npmToken": "<encrypted token>"
},
"npmrc": "registry=https://nexus.example.com/repository/npm-public\n//nexus.example.com/repository/npm-public:_authToken=${NPM_TOKEN}",
 "hostRules": [
  {
    "hostType": "npm",
    "matchHost": "https://nexus.example.com",
    "username": "<plaintext username>",
    "encrypted": {
      "password": "<encrypted password>"
    },
    "authType": "Basic"
  }
],
"privateKey": "$PRIVATE_ENCRYPTION_KEY",
```
From the logs, the decryption seems to work at least for the npm token:
```
DEBUG: Migrating npmToken to npmrc (repository=myuser/exampleproject)
       "decryptedToken": "AB********************************C="
```
Since the revealed characters match.
The password and authentication settings seem to be picked up on also:
```
Found encrypted config (repository=myuser/exampleproject)
       "config": {"password": "***********"}
Trying to decrypt password (repository=myuser/exampleproject)
Could not decrypt using default padding (repository=myuser/exampleproject))
Decrypted password (repository=myuser/exampleproject)
...
Adding password authentication for https://nexus.example.com to hostRules (repository=myuser/exampleproject)
```
But later in the log, renovate fails to authenticate to look up the dependency:
```
Dependency lookup failure: unauthorized (repository=myuser/exampleproject)
"packageUrl": "https://nexus.example.com/repository/npm-public/csv-parser",
"err": {
  "name": "HTTPError",
  code": "ERR_NON_2XX_3XX_RESPONSE",
  },
  "message": "Response code 401 (Unauthorized)",
  "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
  "options": {
    "headers": {
      "user-agent": "RenovateBot/31.65.1 (https://github.com/renovatebot/renovate)",
      "authorization": "***********",
      "accept": "application/json",
      "accept-encoding": "gzip, deflate, br"
    },
    "url": "https://nexus.example.com/repository/npm-public/csv-parser",
    "hostType": "npm",
    "username": "",
    "password": "",
    "method": "GET",
    "http2": false
  },
  "response": {
    "statusCode": 401,
    "statusMessage": "Unauthorized",
    "body": "",
    "headers": {
      "date": "Thu, 10 Feb 2022 11:03:00 GMT",
     "content-length": "0",
      "connection": "close",
      "x-content-type-options": "nosniff",
      "www-authenticate": "BASIC realm=\"Sonatype Nexus Repository Manager\"",
      "strict-transport-security": "max-age=15724800; includeSubDomains"
    },
    "httpVersion": "1.1"
  }
},
"statusCode": 401,
"packageName": "csv-parser"
Failed to look up dependency csv-parser (repository=myuser/exampleproject, packageFile=app/package.json, dependency=csv-parser)
```
The empty `username` and `password` make it look like those settings are not properly applied to the request. If they are used in the `authorization` header, something else must be wrong.
If I manually send a GET request to https://nexus.example.com/repository/npm-public/csv-parser and set the `authorization` header to `Basic <Base64-encoded username:password>`, I get the expected response. So it looks like renovate is not setting the proper header value. Any idea what might be wrong with my configuration?

**Comment by rarkins** [comment]
Why are you configuring both a token as well as a username/password for the same Nexus host?

**Reply by luk-lo** [reply]
The token is just Base64-encoded `username:password` for basic auth, it's not a bearer token if that's what you are asking.
See https://help.sonatype.com/repomanager3/nexus-repository-administration/formats/npm-registry/npm-security#npmSecurity-AuthenticationUsingBasicAuth

**Reply by luk-lo** [reply]
I tried this configuration before too, but same result:
```json
"hostRules": [
  {
    "hostType": "npm",
    "matchHost": "https://nexus.example.com",
    "encrypted": {
      "token": "<encrypted token>"
    },
    "authType": "Basic"
  }
```

**Reply by rarkins** [reply]
My question is why you configured *both*. I think you should focus on `npmrc` and `npmToken` and skip a host rule if it's only duplicating the same credentials.

**Reply by luk-lo** [reply]
I understood from the documentation at https://docs.renovatebot.com/getting-started/private-packages/#dependency-version-lookups that the authentication configuration for the version lookup and the authentication configuration of the package manager at https://docs.renovatebot.com/getting-started/private-packages/#package-manager-credentials-for-artifact-updating are separate things. I did not think the version lookup would know to use basic auth from the npm token configuration.

Removing the `hostRule` and leaving the `npmToken` and `npmrc` items from the configuration did not change the result. Renovate still fails to authenticate when looking up dependencies.

**Reply by viceice** [reply]
`npmToken` is for token only, that doesn't support basic auth tokens! so Please try my suggestions below. Those should work for you, as renovate can use those host rules for update fetch and to run lockfile update.

@rarkins we probably should note this in docs / deprecate the `npmToken` in favor of host rules.

**Comment by viceice** [comment, marked as answer]
Use token or username and password like this:

```json
"npmrc": "registry=https://nexus.example.com/repository/npm-public/",
 "hostRules": [
  {
    "hostType": "npm",
    "matchHost": "https://nexus.example.com/repository/npm-public/",
    "username": "<plaintext username>",
    "encrypted": {
      "password": "<encrypted password>"
    }
  }
]
```

or this:
```json
"npmrc": "registry=https://nexus.example.com/repository/npm-public/",
 "hostRules": [
  {
    "hostType": "npm",
    "matchHost": "https://nexus.example.com/repository/npm-public/",
    "encrypted": {
      "token": "<encrypted token>"
    },
    "authType": "Basic"
  }
]
```

Renovate will add the correspondiung auth line to `.npmrc` for you. Also make sure the urls match.

**Reply by luk-lo** [reply]
Both suggestions did not change the result, though the second one curiously caused the detailed
```
Dependency lookup failure: unauthorized (repository=myuser/exampleproject)
"packageUrl": "https://nexus.example.com/repository/npm-public/csv-parser",
"err": {...},
"statusCode": 401,
"packageName": "csv-parser"
```
part to disappear and only the 
```
Failed to look up dependency csv-parser (repository=myuser/exampleproject, packageFile=app/package.json, dependency=csv-parser)
```
message was output in the log.

**Reply by luk-lo** [reply]
Since the second configuration makes the `unauthorized` part of the error disappear, I'm assuming authentication is working at least. Thank you very much for helping with this.

**Reply by enys** [reply]
did you get this to work ? Seem to end up in the same place, auth works but still 'Failed to lookup'

**Reply by ahached** [reply]
Hello there I'm facing the same issue, where I end up with the 'Failed to lookup', did someone figure out what was the issue ? :/

**Comment by mbmx10** [comment]
I just ran across this question and I am surprised to read about the combination of hostRules and npmrc for self hosted renovate.
The [docs](https://docs.renovatebot.com/getting-started/private-packages/) says about self hosting: 

_Self-hosted hostRules: Configure a hostRules entry in the bot's config.js with the `hostType`, `matchHost ` and `token` specified_,
while `npmrc` is recommended only for The Mend Renovate App. Could it be promising to add the npmrc part in case of occational failed lookups?
